### PR TITLE
Idrac_reset enhancements

### PR DESCRIPTION
**Pull request by singh-lovepreet1**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# ISSUE TYPE

- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- Test Pull Request

##### OUTPUT

```paste below

```
##### ADDITIONAL INFORMATION

```paste below

```
# Checklist:

- [ ] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have maintained backward compatibility

### CODE REVIEW

**Comment by ansible-collections-svc** [comment]
Can one of the admins verify this patch?

File: plugins/modules/idrac_reset.py [hunk 1]
```diff
@@ -3,64 +3,139 @@
...
+version_added: "9.2.0"
 description:
-    - This module resets iDRAC.
-    - "iDRAC is not accessible for some time after running this module. It is recommended to wait for some time,
-    before trying to connect to iDRAC."
+  - This module is responsible to reset the iDRAC to factory default settings.
 extends_documentation_fragment:
   - dellemc.openmanage.idrac_auth_options
+options:
+  reset_to_default:
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/idrac_reset.py:25]: Add version added in documentation of module

File: plugins/modules/idrac_reset.py [hunk 2]
```diff
@@ -3,64 +3,139 @@
...
+      - This parameter provides the option to force reset the iDRAC without checking the iDRAC lifecycle controller status.
+      - This option is applicable only for iDRAC9.
+    type: bool
+    default: false
 
 requirements:
-  - "omsdk >= 1.2.488"
   - "python >= 3.9.6"
 author:
-  - "Felix Stephen (@felixs88)"
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/idrac_reset.py:32]: Don't remove previous authors

File: plugins/modules/idrac_reset.py [hunk 3]
```diff
@@ -92,42 +167,253 @@
...
-        idrac.use_redfish = True
-        msg = idrac.config_mgr.reset_idrac()
-    return msg
+SYSTEMS_URI = "/redfish/v1/Systems"
+MANAGERS_URI = "/redfish/v1/Managers"
+IDRAC_RESET_RETRIES = 10
+LC_STATUS_CHECK_SLEEP = 30
+IDRAC_RESET_OPTIONS_URI = "/redfish/v1/Managers/iDRAC.Embedded.1"
+iDRAC_JOB_URI = "/redfish/v1/Managers/iDRAC.Embedded.1/Jobs/{job_id}"
+IDRAC_RESET_LIFECYCLE_STATUS_URI =  "/redfish/v1/Managers/iDRAC.Embedded.1/Oem/Dell/DellLCService/Actions/DellLCService.GetRemoteServicesAPIStatus"
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/idrac_reset.py]: Check if possible, make URI fetching dynamic, don't hardcode the values

File: plugins/modules/idrac_reset.py [hunk 4]
```diff
@@ -92,42 +167,253 @@
...
+              msg = IDRAC_RESET_SUCCESS_MSG
+              track_failed = False
+              break
+          except URLError as err:
+              time.sleep(interval)
+              wait = wait - interval
+      return track_failed, status_code, msg
+             
+    def __reset_to_default_all(self):
+        payload = {}
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/idrac_reset.py]: Remove all these functions and keep a single function which consume different parameters, as only reset_type is changing, reset all code is same. Will be easy to maintain in future.

File: plugins/modules/idrac_reset.py (hunk 1 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py:25]: Done!

File: plugins/modules/idrac_reset.py (hunk 2 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py:32]: Done!

File: plugins/modules/idrac_reset.py (hunk 4 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Done!

File: plugins/modules/idrac_reset.py [hunk 5]
```diff
@@ -3,64 +3,141 @@
...
 
 requirements:
-  - "omsdk >= 1.2.488"
   - "python >= 3.9.6"
 author:
   - "Felix Stephen (@felixs88)"
   - "Anooja Vardhineni (@anooja-vardhineni)"
+  - "Lovepreet Singh (@singh-lovepreet1)"
 notes:
-    - Run this module from a system that has direct access to Dell iDRAC.
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py:35]: Don't remove this note.

File: plugins/modules/idrac_reset.py [hunk 6]
```diff
@@ -3,64 +3,141 @@
...
-EXAMPLES = """
+EXAMPLES = r'''
 ---
-- name: Reset iDRAC
+- name: Reset the iDRAC to all and wait for the iDRAC to be up.
+  dellemc.openmanage.idrac_reset:
+   idrac_ip: "192.168.0.1"
+   idrac_user: "user_name"
+   idrac_password: "user_password"
+   reset_to_default: "All"
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py:92]: Add ca_path: "/path/to/ca_cert.pem" in all example playbooks.

File: plugins/modules/idrac_reset.py [hunk 7]
```diff
@@ -3,64 +3,141 @@
...
+'''
 
 RETURN = r'''
 ---
 msg:
-  description: Status of the iDRAC reset operation.
-  returned: always
   type: str
-  sample: "Successfully performed iDRAC reset."
+  description: Overall status of the storage configuration operation.
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py]: Status of the iDRAC reset operation.

File: plugins/modules/idrac_reset.py [hunk 8]
```diff
@@ -3,64 +3,141 @@
...
+  returned: always
+  sample: "Successfully completed the "
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py]: This is incomplete. Add the appropriate message.

File: plugins/modules/idrac_reset.py [hunk 9]
```diff
@@ -92,41 +169,312 @@
...
-from ansible.module_utils.six.moves.urllib.error import URLError, HTTPError
+from ansible.module_utils.compat.version import LooseVersion
 from ansible.module_utils.urls import ConnectionError, SSLValidationError
+from ansible_collections.dellemc.openmanage.plugins.module_utils.utils import (
+    get_idrac_firmware_version, remove_key, idrac_redfish_job_tracking)
+
+
+SYSTEMS_URI = "/redfish/v1/Systems"
+MANAGERS_URI = "/redfish/v1/Managers"
+iDRAC_JOB_URI = "/redfish/v1/Managers/iDRAC.Embedded.1/Jobs/{job_id}"
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py]: Change it to IDRAC_JOB_URI

File: plugins/modules/idrac_reset.py [hunk 10]
```diff
@@ -92,41 +169,312 @@
...
+IDRAC_RESET_RETRIES = 10
+LC_STATUS_CHECK_SLEEP = 30
+IDRAC_RESET_OPTIONS_URI = "/redfish/v1/Managers/iDRAC.Embedded.1"
+iDRAC_JOB_URI = "/redfish/v1/Managers/iDRAC.Embedded.1/Jobs/{job_id}"
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py]: Remove this. Already defined above.

File: plugins/modules/idrac_reset.py [hunk 11]
```diff
@@ -92,41 +169,312 @@
...
+IDRAC_RESET_LIFECYCLE_STATUS_URI = "/redfish/v1/Managers/iDRAC.Embedded.1/Oem/Dell/DellLCService/Actions/DellLCService.GetRemoteServicesAPIStatus"
+IDRAC_RESET_SET_CUSTOM_DEFAULTS_URI = "/redfish/v1/Managers/{ManagerId}/Actions/Oem/DellManager.SetCustomDefaults"
+IDRAC_RESET_GET_CUSTOM_DEFAULTS_URI = "/redfish/v1/Managers/iDRAC.Embedded.1/Oem/Dell/CustomDefaultsDownloadURI"
+IDRAC_RESET_GRACEFUL_RESTART_URI = "/redfish/v1/Managers/iDRAC.Embedded.1/Actions/Manager.Reset"
+IDRAC_RESET_RESET_URI = "/redfish/v1/Managers/iDRAC.Embedded.1/Actions/Oem/DellManager.ResetToDefaults"
+RESET_TO_DEFAULT_ERROR = "{0} is not supported. The supported values are {1}. Enter the valid values and retry the operation."
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py]: Instead of using 0 and 1 in formatting, use appropriate names.

File: plugins/modules/idrac_reset.py [hunk 12]
```diff
@@ -92,41 +169,312 @@
...
+IDRAC_RESET_RESTART_SUCCESS_MSG = "iDRAC restart operation completed successfully"
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py]: Few msgs have full stops. Few don't. Add full stops in all the message variables.

File: plugins/modules/idrac_reset.py [hunk 13]
```diff
@@ -92,41 +169,312 @@
...
+IDRAC_RESET_SUCCESS_MSG = "Successfully performed iDRAC reset."
+IDRAC_RESET_RESET_TRIGGER_MSG = "iDRAC reset operation triggered successfully"
+IDRAC_RESET_RESTART_TRIGGER_MSG = "iDRAC restart operation triggered successfully"
+INVALID_VALUE_MSG = "The value for the `{parameter}` parameter is invalid."
+NOT_ALLOWED_VALUE_MSG = "`{reset_to_default}` is not supported. The supported values are `{supported_reset_to_default_values}`. \
+                          Enter the valid values and retry the operation."
+WAIT_TIMEOUT_MSG = "The job is not completed after {0} seconds."
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py]: Instead of using 0  in formatting, use appropriate names.

File: plugins/modules/idrac_reset.py [hunk 14]
```diff
@@ -92,41 +169,312 @@
...
+            upload_perfom = True
+            default_data = custom_default_buffer
+        if upload_perfom:
+            self.upload_cd_file(default_data)
+        self.validate_obj.validate_custom_option(self.reset_to_default, self.allowed_choices)
+        return self.reset_to_default_mapped()
+
+    def graceful_restart(self):
+        payload = {}
+        payload["ResetType"] = "GracefulRestart"
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py]: Define this in one line: 
 payload = {"ResetType": "GracefulRestart"}

File: plugins/modules/idrac_reset.py [hunk 15]
```diff
@@ -92,41 +169,312 @@
...
+                    track_failed = False
+                    break
+            except URLError as err:
+                time.sleep(interval)
+                wait = wait - interval
+        return track_failed, status_code, msg
+
+    def reset_to_default_mapped(self):
+        payload = {}
+        payload["ResetType"] = self.reset_to_default
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py]: Directly use - payload = {"ResetType": self.reset_to_default}

File: plugins/modules/idrac_reset.py [hunk 16]
```diff
@@ -92,41 +169,312 @@
...
+        self.validate_obj.validate_reset_options(self.allowed_choices, RESET_KEY)
+        return self.perform_operation(payload)
+
+    def reset_custom_defaults(self):
+        # add a check for firmware version for this > 7.00.30.00
+        if self.idrac_firmware_version < '7.00.30.00':
+            self.module.exit_json(msg=RESET_TO_DEFAULT_ERROR.format(self.reset_to_default, self.allowed_choices), skipped=True)
+        tmp_res, res = None, None
```
Comment (Kritika-Bhateja-03) [inline plugins/modules/idrac_reset.py]: Not used anywhere. Remove these.

Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Done

File: plugins/modules/idrac_reset.py (hunk 15 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Done

File: plugins/modules/idrac_reset.py (hunk 14 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Done

File: plugins/modules/idrac_reset.py (hunk 13 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Updated!

File: plugins/modules/idrac_reset.py (hunk 12 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Updated!

File: plugins/modules/idrac_reset.py (hunk 11 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Updated!

File: plugins/modules/idrac_reset.py (hunk 10 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Done!

File: plugins/modules/idrac_reset.py (hunk 9 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Done

File: plugins/modules/idrac_reset.py (hunk 8 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Updated

File: plugins/modules/idrac_reset.py (hunk 7 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Updated

File: plugins/modules/idrac_reset.py (hunk 6 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py:92]: Updated

File: plugins/modules/idrac_reset.py (hunk 5 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py:35]: Added

File: plugins/modules/idrac_reset.py (hunk 3 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Updated!

File: plugins/modules/idrac_reset.py [hunk 17]
```diff
@@ -92,41 +175,340 @@
...
-from ansible.module_utils.six.moves.urllib.error import URLError, HTTPError
+from ansible.module_utils.compat.version import LooseVersion
 from ansible.module_utils.urls import ConnectionError, SSLValidationError
+from ansible_collections.dellemc.openmanage.plugins.module_utils.utils import (
+    get_idrac_firmware_version, remove_key, get_dynamic_uri, validate_and_get_first_resource_id_uri, idrac_redfish_job_tracking)
+
+
+MANAGERS_URI = "/redfish/v1/Managers"
+OEM = "Oem"
+MANUFACTURER = "Dell"
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/idrac_reset.py:197]: This is not proper way to do dynamic uri fetching, here you are hardcoding url in multiple variables and using same below. 

In dynamic uri fetching, you should have only base uri and then reach to target uri dynamically. It'll helps in future if any url changes. It won't break the code.

Please refer to module like idrac_network_attributes, where uri is being fetched dynamically.

Comment (ABHISHEK-SINHA10) [reply plugins/modules/idrac_reset.py:197]: It's fetching dynamically, but one suggestion, only keep variables of those which is being used in more than one place like ACTIONS, OEM etc., if any variable used in single place, put it directly in code, no need to create variable for single use.

Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py:197]: Updated accordingly

File: plugins/modules/idrac_reset.py [hunk 18]
```diff
@@ -3,64 +3,147 @@
...
+      - This parameter provides the option to import the buffer input in XML format as a custom default configuration.
+      - This option is applicable when I(reset_to_default) is C(CustomDefaults).
+      - I(custom_defaults_buffer) is mutually exclusive with I(custom_defaults_file).
+    type: str
+  wait_for_idrac:
+    description:
+      - This parameter provides the option to wait for the iDRAC to reset and lifecycle controller status to be ready.
+    type: bool
+    default: true
+  job_wait_timeout:
```
Comment (ABHISHEK-SINHA10) [inline plugins/modules/idrac_reset.py:55]: Job_wait_timeout is not in BSpec, Please check once and update.

File: plugins/modules/idrac_reset.py [hunk 19]
```diff
@@ -3,64 +3,147 @@
...
 __metaclass__ = type
 
-DOCUMENTATION = """
+DOCUMENTATION = r'''
 ---
 module: idrac_reset
-short_description: Reset iDRAC
-version_added: "2.1.0"
+short_description: Factory reset the iDRACs
+version_added: "9.2.0"
```
Comment (sachin-apa) [inline plugins/modules/idrac_reset.py]: Keep this as 2.1.0, it was added at that releases

**Review by sachin-apa — COMMENTED** [review]
Lets add documentation as well

File: plugins/modules/idrac_reset.py (hunk 19 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py]: Done

File: plugins/modules/idrac_reset.py (hunk 18 above)
Comment (singh-lovepreet1) [reply plugins/modules/idrac_reset.py:55]: Added a comment for updating Bspec

**Comment by singh-lovepreet1** [comment]
UT COVERAGE:

image

**Review by sachin-apa — DISMISSED** [review]

**Review by sachin-apa — APPROVED** [review]

**Review by rajshekarp87 — APPROVED** [review]

**Review by jagadeeshnv — APPROVED** [review]

**Review by ShivamSh3 — APPROVED** [review]

**Pull request merged**